File: Solutions/Python/combinations.py
"""
    [MEDIUM]
    77. Combinations

    Concepts:
    - backtracking

    Stats:
        Runtime | 147 ms    [Beats 35.61%]
        Memory  | 59.60 MB  [Beats 48.60%]
"""

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []

        def dfs(num: int, path: list[int]):
            nonlocal res

            if len(path) == k:
                res.append(path)
                return

            for i in range(num, n + 1):
                dfs(i + 1, path + [i])

        dfs(1, [])
        return res

File: Solutions/Cpp/combinations.cpp
/*
    [MEDIUM]
    77. Combinations

    Concepts:
    - backtracking

    Stats:
        Runtime | 50 ms     [Beats 48.13%]
        Memory  | 63.73 MB  [Beats 47.67%]
*/

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        std::vector<std::vector<int>> res = {};
        std::vector<int> path = {};

        std::function<void(int)> dfs;
        dfs = [&](int num){
            if (path.size() == k)
                return res.push_back(path);

            for (int i = num; i <= n; i++) {
                path.push_back(i);
                dfs(i + 1);
                path.pop_back();
            }
        };

        dfs(1);
        return res;
    }
};